### PR TITLE
[compiler][api] Enable custom registration of Relay ops from python

### PR DESCRIPTION
**Pull request by bwasti**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42 [compiler][api] Enable custom registration of Relay ops from python**

### CODE REVIEW

File: torch_tvm/register.cpp [hunk 1]
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
...
+            auto op = tvm::relay::Op::Get(relay_name);
+            AT_ASSERT(op);
+            return tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
+          }}});
+
+    auto relay_op = tvm::relay::Op::Get(relay_name);
+    AT_ASSERT(relay_op);
+    std::vector<Argument> torch_inputs;
+    Graph wrapper_graph;
+    std::vector<Value *> graph_inputs;
```
Comment (wanchaol) [inline torch_tvm/register.cpp]: nit: `Value*`?

File: torch_tvm/register.cpp [hunk 2]
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
...
+    RegisterTVMOperator reg_tvm_operator(
+        {{Symbol::fromQualString("tvm::" + torch_name),
+          [relay_name](Node *node, tvm::Array<tvm::relay::Expr> inputs) {
+            auto op = tvm::relay::Op::Get(relay_name);
+            AT_ASSERT(op);
+            return tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
+          }}});
+
+    auto relay_op = tvm::relay::Op::Get(relay_name);
+    AT_ASSERT(relay_op);
```
Comment (wanchaol) [inline torch_tvm/register.cpp]: this should be `AT_CHECK` and send msg to user if there's no such op exist?

File: torch_tvm/register.cpp [hunk 3]
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
...
 
   m.def("disable", []() { fusion_enabled = false; });
+  m.def("register_operator", [tvm_sym](std::string torch_name,
+                                       std::string relay_name) {
+    RegisterTVMOperator reg_tvm_operator(
+        {{Symbol::fromQualString("tvm::" + torch_name),
+          [relay_name](Node *node, tvm::Array<tvm::relay::Expr> inputs) {
+            auto op = tvm::relay::Op::Get(relay_name);
+            AT_ASSERT(op);
+            return tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
```
Comment (wanchaol) [inline torch_tvm/register.cpp]: does this mean that we currently can only register simple operators that does not require custom Attrs?

File: torch_tvm/register.cpp [hunk 4]
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
...
+    auto sym = Symbol::fromQualString("tvm::" + torch_name);
+    Node *node = wrapper_graph.create(sym, graph_inputs);
+    wrapper_graph.appendNode(node);
+    AT_CHECK(node->outputs().size() == 1,
+             "Currently only single output relay ops "
+             "can be custom registered from Python.");
+    wrapper_graph.registerOutput(node->output());
+    // Convert node to a tvm compilation group containing the node
+    node = SubgraphUtils::createSingletonSubgraph(node, tvm_sym);
+    auto cc = std::make_shared<TVMCompiler>(node, opt_level, strict,
```
Comment (wanchaol) [inline torch_tvm/register.cpp]: from what I observed here, for each custom operators, we are creating a single separate `TVMCompiler` group and execute them separately? will the `customFusion` pass take the compilation group and merge it with the bigger one?

File: test/test_integration.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+        @torch.jit.script
+        def test(a, b):
+          d = torch.ops.tvm.woo(a)
+          e = torch.ops.tvm.woo(b)
+          return d + e
+        
+        X = torch.randn(128,128)
+        print(test.graph)
+        print(test.graph_for(X,X))
```
Comment (wanchaol) [inline test/test_integration.py]: it is hard to see the graph from the PR, why don't we just call `runBoth` here?

File: torch_tvm/register.cpp (hunk 3 above)
Comment (bwasti) [reply torch_tvm/register.cpp]: yep, unfortunately.  There isn't a great way to expose the relay Attrs interface to python (as far as I can tell)

File: torch_tvm/register.cpp (hunk 4 above)
Comment (bwasti) [reply torch_tvm/register.cpp]: customFusion will never be run on these operators because this interface is for eager mode only

**Review by wanchaol — APPROVED** [review]
looks good.

**Comment by bwasti** [comment]
There are some issues with this API that make it pretty useless, so I am going to abandon it.  That being said, there are some tricks here worth using

**Pull request closed without merging**